File: PhotoRefinementCoach/src/trait_to_prompt.py
"""
Trait-to-Prompt Converter

Converts structured PaDNA trait data into optimized prompts for image generation.
Supports multiple rendering engines (Stable Diffusion, FLUX, Midjourney, etc.)
"""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

logger = logging.getLogger(__name__)


class TraitCategory:
    """Represents a category of traits with rendering priority."""

    def __init__(
        self,
        name: str,
        priority: int,
        prefix: str,
        weight: float = 1.0
    ):
        self.name = name
        self.priority = priority  # Lower = higher priority
        self.prefix = prefix
        self.weight = weight  # Emphasis weight in prompt


# Trait category configuration
TRAIT_CATEGORIES = {
    "face": TraitCategory("Face", 1, "PaDNA.FacialDNA", 1.2),
    "eyes": TraitCategory("Eyes", 2, "PaDNA.EyeDNA", 1.1),
    "hair": TraitCategory("Hair", 3, "PaDNA.HairDNA", 1.1),
    "skin": TraitCategory("Skin", 4, "PaDNA.SkinDNA", 1.0),
    "smile": TraitCategory("Smile", 5, "PaDNA.SmileDNA", 0.9),
    "body": TraitCategory("Body", 6, "PaDNA.BodyDNA", 0.8),
    "expression": TraitCategory("Expression", 7, "PaDNA.ExpressionDNA", 0.7),
    "fitness": TraitCategory("Fitness", 8, "PaDNA.FitnessDNA", 0.6),
    "apparel": TraitCategory("Apparel", 9, "PaDNA.ApparelDNA", 0.5),
}


class PromptBuilder:
    """Builds optimized prompts from PaDNA traits."""

    def __init__(self, style: str = "photorealistic"):
        """
        Initialize prompt builder.

        Args:
            style: Rendering style (photorealistic, portrait, cinematic, artistic)
        """
        self.style = style
        self.trait_order = self._build_trait_order()

    def _build_trait_order(self) -> List[str]:
        """Define the order traits should appear in prompts (most important first)."""
        return [
            # Core appearance (highest priority)
            "PaDNA.EyeDNA.Color",
            "PaDNA.EyeDNA.Shape",
            "PaDNA.HairDNA.Color",
            "PaDNA.HairDNA.Length",
            "PaDNA.HairDNA.Texture",
            "PaDNA.SkinDNA.Tone",
            "PaDNA.SkinDNA.Freckles",

            # Facial features
            "PaDNA.SmileDNA.SmileShape",
            "PaDNA.SmileDNA.Teeth",
            "PaDNA.EyeDNA.Lashes",
            "PaDNA.FacialDNA.FaceShape",
            "PaDNA.HairDNA.Parting",
            "PaDNA.HairDNA.Volume",

            # Expression
            "PaDNA.ExpressionDNA.TypicalExpression",
            "PaDNA.ExpressionDNA.EyeContact",

            # Body (secondary)
            "PaDNA.BodyDNA.Build",
            "PaDNA.BodyDNA.HeightEstimateCM",
            "PaDNA.BodyDNA.WaistHipRatio",
            "PaDNA.BodyDNA.Legs",
            "PaDNA.BodyDNA.Arms",
            "PaDNA.BodyDNA.Bust",
            "PaDNA.BodyDNA.Shoulders",
            "PaDNA.FitnessDNA.Indicators",

            # Details
            "PaDNA.BodyDNA.Navel",
            "PaDNA.ApparelDNA.Jewelry.Necklace",
            "PaDNA.ApparelDNA.Jewelry.Earrings",
            "PaDNA.ApparelDNA.StyleThemes",
        ]

    def build_prompt(
        self,
        traits: Dict[str, Dict[str, Any]],
        emphasis_threshold: float = 800.0,
        include_style_prefix: bool = True,
        max_tokens: int = 200
    ) -> str:
        """
        Build a natural language prompt from traits.

        Args:
            traits: Dictionary of trait paths to trait data
            emphasis_threshold: UCN threshold for emphasized traits
            include_style_prefix: Include style description at start
            max_tokens: Approximate max tokens (words) in prompt

        Returns:
            Optimized prompt string
        """
        segments = []

        # Style prefix
        if include_style_prefix:
            style_prefixes = {
                "photorealistic": "Professional high-quality portrait photograph, photorealistic, 8k, detailed,",
                "portrait": "Studio portrait, professional lighting, detailed features,",
                "cinematic": "Cinematic portrait, film quality, dramatic lighting, detailed,",
                "artistic": "Artistic portrait, elegant composition, refined details,",
            }
            segments.append(style_prefixes.get(self.style, style_prefixes["photorealistic"]))

        # Build descriptions by category priority
        descriptions = self._extract_descriptions(traits, emphasis_threshold)

        # Add high-priority features first
        face_parts = []

        # Eyes
        eye_desc = descriptions.get("eyes", [])
        if eye_desc:
            face_parts.extend(eye_desc)

        # Hair
        hair_desc = descriptions.get("hair", [])
        if hair_desc:
            face_parts.extend(hair_desc)

        # Skin
        skin_desc = descriptions.get("skin", [])
        if skin_desc:
            face_parts.extend(skin_desc)

        # Face features
        face_desc = descriptions.get("face", [])
        smile_desc = descriptions.get("smile", [])
        if smile_desc:
            face_parts.extend(smile_desc)
        if face_desc:
            face_parts.extend(face_desc)

        # Core subject with face features
        if face_parts:
            segments.append("a woman with " + ", ".join(face_parts) + ".")
        else:
            segments.append("a woman.")

        # Expression
        expression_desc = descriptions.get("expression", [])
        if expression_desc:
            segments.append("Expression: " + ", ".join(expression_desc[:2]) + ".")

        # Body and fitness
        body_parts = []
        body_desc = descriptions.get("body", [])
        fitness_desc = descriptions.get("fitness", [])

        if body_desc:
            body_parts.extend(body_desc[:4])  # Limit body details
        if fitness_desc:
            body_parts.extend(fitness_desc[:2])  # Limit fitness details

        if body_parts:
            segments.append("Body: " + ", ".join(body_parts) + ".")

        # Details (jewelry, piercings)
        apparel_desc = descriptions.get("apparel", [])
        if apparel_desc:
            details = apparel_desc[:3]  # Limit to key details
            if details:
                segments.append("Wearing: " + ", ".join(details) + ".")

        # Join segments
        prompt = " ".join(segments)

        # Add quality keywords
        prompt += ", sharp focus, high detail, natural lighting"

        # Truncate if too long (approximate)
        words = prompt.split()
        if len(words) > max_tokens:
            prompt = " ".join(words[:max_tokens]) + "..."

        return prompt

    def _extract_descriptions(
        self,
        traits: Dict[str, Dict[str, Any]],
        emphasis_threshold: float
    ) -> Dict[str, List[str]]:
        """
        Extract natural language descriptions organized by category.

        Args:
            traits: Trait dictionary
            emphasis_threshold: UCN threshold for emphasis

        Returns:
            Dictionary of category -> list of descriptions
        """
        descriptions: Dict[str, List[str]] = {}

        for trait_path, trait_data in traits.items():
            if not isinstance(trait_data, dict):
                continue

            value = trait_data.get("resolved_value")
            ucn = trait_data.get("ucn", 0)

            if value is None:
                continue

            # Determine category
            category = self._categorize_trait(trait_path)
            if category not in descriptions:
                descriptions[category] = []

            # Convert trait to description
            desc = self._trait_to_description(trait_path, value, ucn, emphasis_threshold)
            if desc:
                descriptions[category].append(desc)

        return descriptions

    def _categorize_trait(self, path: str) -> str:
        """Determine category for a trait path."""
        for cat_key, category in TRAIT_CATEGORIES.items():
            if path.startswith(category.prefix):
                return cat_key
        return "other"

    def _trait_to_description(
        self,
        path: str,
        value: Any,
        ucn: float,
        emphasis_threshold: float
    ) -> Optional[str]:
        """
        Convert a single trait to a natural language description.

        Args:
            path: Trait path
            value: Trait value
            ucn: Confidence score
            emphasis_threshold: UCN threshold for emphasis

        Returns:
            Natural language description or None
        """
        # Handle list values
        if isinstance(value, list):
            if not value:
                return None
            # Join list items
            if len(value) == 1:
                value = str(value[0])
            else:
                value = ", ".join(str(v) for v in value[:3])  # Limit to 3 items

        # Convert to string for processing
        if isinstance(value, str):
            value_display = value
            value_str = value.lower()
        else:
            value_display = str(value)
            value_str = str(value).lower()

        # Skip certain values
        skip_values = ["absent", "none", "n/a", "unknown"]
        if value_str in skip_values:
            return None

        # Map trait paths to natural descriptions
        descriptions = {
            # Eyes
            "PaDNA.EyeDNA.Color": f"{value_display} eyes",
            "PaDNA.EyeDNA.Shape": f"{value_str}-shaped eyes",
            "PaDNA.EyeDNA.Lashes": f"{value_str} eyelashes",
            "PaDNA.EyeDNA.EyeSpacing": f"{value_str} eye spacing",

            # Hair
            "PaDNA.HairDNA.Color": f"{value_str} hair",
            "PaDNA.HairDNA.Length": f"{value_str} hair",
            "PaDNA.HairDNA.Texture": f"{value_str} hair",
            "PaDNA.HairDNA.Volume": f"{value_str} voluminous hair",
            "PaDNA.HairDNA.Parting": f"hair with {value_str}",

            # Skin
            "PaDNA.SkinDNA.Tone": f"{value_str} skin",
            "PaDNA.SkinDNA.Freckles": f"{value_str} freckles" if "present" in value_str else None,

            # Face/Smile
            "PaDNA.SmileDNA.SmileShape": f"{value_str}",
            "PaDNA.SmileDNA.Teeth": f"{value_str} teeth",
            "PaDNA.SmileDNA.Dimples": f"dimples" if "present" in value_str else None,

            # Expression
            "PaDNA.ExpressionDNA.TypicalExpression": f"{value_str}",
            "PaDNA.ExpressionDNA.EyeContact": f"{value_str}",

            # Body
            "PaDNA.BodyDNA.Build": f"{value_str} build",
            "PaDNA.BodyDNA.WaistHipRatio": f"{value_str} figure",
            "PaDNA.BodyDNA.Legs": f"{value_str} legs",
            "PaDNA.BodyDNA.Arms": f"{value_str} arms",
            "PaDNA.BodyDNA.Bust": f"{value_str} bust",
            "PaDNA.BodyDNA.Navel": "navel piercing" if "pierced" in value_str else None,

            # Fitness
            "PaDNA.FitnessDNA.Indicators": f"{value_str}",

            # Apparel/Jewelry
            "PaDNA.ApparelDNA.Jewelry.Necklace": f"wearing {value_str}",
            "PaDNA.ApparelDNA.Jewelry.Earrings": f"wearing {value_str}",
        }

        desc = descriptions.get(path)

        # Fallback: extract last part of path and use value
        if desc is None and value:
            trait_name = path.split(".")[-1].lower()
            desc = f"{value_str}"

        # Add emphasis for high-confidence traits
        if desc and ucn >= emphasis_threshold:
            # Could add parentheses or other emphasis markers here
            pass

        return desc


class ImageGenerationRequest:
    """Represents a request for image generation."""

    def __init__(
        self,
        prompt: str,
        negative_prompt: str = "",
        style: str = "photorealistic",
        width: int = 768,
        height: int = 1024,
        steps: int = 30,
        cfg_scale: float = 7.0,
        seed: Optional[int] = None,
    ):
        self.prompt = prompt
        self.negative_prompt = negative_prompt or self._default_negative_prompt()
        self.style = style
        self.width = width
        self.height = height
        self.steps = steps
        self.cfg_scale = cfg_scale
        self.seed = seed

    def _default_negative_prompt(self) -> str:
        """Default negative prompt to avoid common issues."""
        return (
            "blurry, low quality, distorted, deformed, ugly, bad anatomy, "
            "bad proportions, extra limbs, mutated, disfigured, "
            "watermark, text, signature, logo, multiple heads, duplicate, "
            "cartoon, anime, painting, illustration"
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for API calls."""
        return {
            "prompt": self.prompt,
            "negative_prompt": self.negative_prompt,
            "width": self.width,
            "height": self.height,
            "steps": self.steps,
            "cfg_scale": self.cfg_scale,
            "seed": self.seed,
        }


def generate_prompt_from_resolved(
    resolved_path: Path,
    style: str = "photorealistic",
    emphasis_threshold: float = 800.0
) -> Tuple[str, ImageGenerationRequest]:
    """
    Generate an image prompt from a resolved.json file.

    Args:
        resolved_path: Path to resolved.json
        style: Rendering style
        emphasis_threshold: UCN threshold for emphasis

    Returns:
        Tuple of (prompt text, ImageGenerationRequest)
    """
    with resolved_path.open("r") as f:
        traits = json.load(f)

    builder = PromptBuilder(style=style)
    prompt = builder.build_prompt(traits, emphasis_threshold=emphasis_threshold)

    request = ImageGenerationRequest(
        prompt=prompt,
        style=style,
        width=768,
        height=1024,
        steps=40,
        cfg_scale=7.5,
    )

    return prompt, request


def generate_prompt_from_traits(
    traits: Dict[str, Dict[str, Any]],
    style: str = "photorealistic",
    emphasis_threshold: float = 800.0
) -> str:
    """
    Generate an image prompt from trait dictionary.

    Args:
        traits: Dictionary of trait paths to trait data
        style: Rendering style
        emphasis_threshold: UCN threshold for emphasis

    Returns:
        Optimized prompt string
    """
    builder = PromptBuilder(style=style)
    return builder.build_prompt(traits, emphasis_threshold=emphasis_threshold)

File: web/src/server/trait-to-prompt.ts
/**
 * Trait-to-Prompt Converter v2.1
 *
 * Converts structured PaDNA trait data into optimized prompts for image generation.
 * TypeScript port of PhotoRefinementCoach/src/trait_to_prompt.py
 *
 * v2.1 Enhancements:
 * - Priority weighting for high-impact containers (freckles, iris, texture, makeup)
 * - Composite descriptors for iris detail and freckles
 * - Enhanced makeup and accessory mappings
 * - Anti-airbrushing negative prompts
 */

export interface Trait {
  resolved_value: any;
  ucn?: number;
  rr?: number;
  [key: string]: any;
}

export type TraitDict = Record<string, Trait>;

export type PromptStyle = 'photorealistic' | 'portrait' | 'cinematic' | 'artistic';

interface TraitDescription {
  [category: string]: string[];
}

// v2.1: Priority weighting for critical realism containers
const PRIORITY_WEIGHTS: Record<string, 'high' | 'med' | 'std'> = {
  'PaDNA.SkinDNA.Texture': 'high',
  'PaDNA.SkinDNA.Freckles.Density': 'high',
  'PaDNA.SkinDNA.Freckles.Coverage': 'high',
  'PaDNA.SkinDNA.Freckles.Contrast': 'high',
  'PaDNA.EyeDNA.Iris.BaseColor': 'high',
  'PaDNA.EyeDNA.Iris.Flecks': 'high',
  'PaDNA.EyeDNA.Iris.FleckPattern': 'high',
  'PaDNA.EyeDNA.Iris.LimbalRing': 'high',
  'MakeupDNA.Eyes.Style': 'med',
  'MakeupDNA.Lips.Finish': 'med',
};

const WEIGHT_VALUES = { high: 2.0, med: 1.5, std: 1.0 };

const STYLE_PREFIXES: Record<PromptStyle, string> = {
  photorealistic: 'Professional high-quality portrait photograph, photorealistic, 8k, detailed,',
  portrait: 'Studio portrait, professional lighting, detailed features,',
  cinematic: 'Cinematic portrait, film quality, dramatic lighting, detailed,',
  artistic: 'Artistic portrait, elegant composition, refined details,',
};

const TRAIT_ORDER = [
  // Core appearance (highest priority)
  'PaDNA.EyeDNA.Color',
  'PaDNA.EyeDNA.Shape',
  'PaDNA.HairDNA.Color',
  'PaDNA.HairDNA.Length',
  'PaDNA.HairDNA.Texture',
  'PaDNA.SkinDNA.Tone',
  'PaDNA.SkinDNA.Freckles',
  // Facial features
  'PaDNA.SmileDNA.SmileShape',
  'PaDNA.SmileDNA.Teeth',
  'PaDNA.EyeDNA.Lashes',
  'PaDNA.HairDNA.Parting',
  'PaDNA.HairDNA.Volume',
  // Expression
  'PaDNA.ExpressionDNA.TypicalExpression',
  'PaDNA.ExpressionDNA.EyeContact',
  // Body
  'PaDNA.BodyDNA.Build',
  'PaDNA.BodyDNA.WaistHipRatio',
  'PaDNA.BodyDNA.Legs',
  'PaDNA.BodyDNA.Arms',
  // Details
  'PaDNA.BodyDNA.Navel',
  'PaDNA.ApparelDNA.Jewelry.Necklace',
  'PaDNA.ApparelDNA.Jewelry.Earrings',
];

function categorizeTraitPath(path: string): string {
  if (path.includes('.EyeDNA.')) return 'eyes';
  if (path.includes('.HairDNA.')) return 'hair';
  if (path.includes('.SkinDNA.')) return 'skin';
  if (path.includes('.SmileDNA.') || path.includes('.FacialDNA.')) return 'face';
  if (path.includes('.ExpressionDNA.')) return 'expression';
  if (path.includes('.BodyDNA.')) return 'body';
  if (path.includes('.FitnessDNA.')) return 'fitness';
  if (path.includes('.ApparelDNA.') || path.includes('.Accessories.')) return 'apparel';
  if (path.includes('MakeupDNA')) return 'makeup';
  return 'other';
}

/**
 * v2.1: Composite descriptor for freckles (combines density + coverage + contrast)
 */
function frecklesComposite(density?: string, coverage?: string, contrast?: string): string | null {
  if (!density || density === 'none') return null;

  const densityMap: Record<string, string> = {
    'light': 'light freckles',
    'medium': 'medium density freckles',
    'heavy': 'heavy freckles',
    'very-heavy': 'very heavy freckles',
  };

  const coverageMap: Record<string, string> = {
    'cheeks-only': 'on cheeks',
    't-zone': 'across nose bridge and t-zone',
    'full-face': 'covering the face',
    'face-shoulders': 'across face, neck and shoulders',
    'face-neck-shoulders': 'across face, neck and shoulders',
  };

  const contrastMap: Record<string, string> = {
    'subtle': 'subtle',
    'visible': 'visible',
    'prominent': 'prominent',
    'stark': 'high-contrast',
  };

  const d = densityMap[density] || `${density} freckles`;
  const cvr = coverage ? coverageMap[coverage] : '';
  const ctr = contrast ? contrastMap[contrast] : '';

  // Build: "prominent medium density freckles across face, neck and shoulders"
  const parts = [ctr, d, cvr].filter(Boolean);
  return parts.join(' ').trim();
}

/**
 * v2.1: Composite descriptor for iris (combines base color + flecks + pattern + limbal ring)
 */
function irisComposite(base?: string, flecks?: string, pattern?: string, ring?: string): string | null {
  if (!base) return null;

  const baseColorMap: Record<string, string> = {
    'light-green': 'striking light green eyes',
    'hazel': 'hazel eyes',
    'amber': 'warm amber eyes',
    'light-blue': 'bright light blue eyes',
    'gray': 'cool gray eyes',
  };

  const patternMap: Record<string, string> = {
    'scattered': ' scattered throughout the iris',
    'central': ' concentrated near the pupil',
    'radial': ' radiating from the pupil',
    'dense': ' densely distributed across the iris',
  };

  const ringMap: Record<string, string> = {
    'faint': ', subtle limbal ring',
    'moderate': ', moderate limbal ring adding depth',
    'prominent': ', prominent limbal ring',
    'dark-prominent': ', dark prominent limbal ring for striking eyes',
  };

  const baseText = baseColorMap[base.toLowerCase()] || `${base.replace(/-/g, ' ')} eyes`;
  const fleckText = (flecks && flecks !== 'none') ? ` with ${flecks} flecks` : '';
  const patternText = (pattern && pattern !== 'none') ? (patternMap[pattern] || '') : '';
  const ringText = (ring && ring !== 'none') ? (ringMap[ring] || '') : '';

  return `${baseText}${fleckText}${patternText}${ringText}`.trim();
}

function traitToDescription(path: string, value: any): string | null {
  // Handle list values
  if (Array.isArray(value)) {
    if (value.length === 0) return null;
    if (value.length === 1) {
      value = String(value[0]);
    } else {
      value = value.slice(0, 3).join(', ');
    }
  }

  // Convert to string
  const valueStr = String(value).toLowerCase();
  const valueDisplay = String(value);

  // Skip certain values
  const skipValues = ['absent', 'none', 'n/a', 'unknown'];
  if (skipValues.includes(valueStr)) {
    return null;
  }

  // Map trait paths to natural descriptions
  const descriptions: Record<string, string | null> = {
    // Eyes
    'PaDNA.EyeDNA.Color': `${valueDisplay} eyes`,
    'PaDNA.EyeDNA.Shape': `${valueStr}-shaped eyes`,
    'PaDNA.EyeDNA.Lashes': `${valueStr} eyelashes`,
    // Eyes - Enhanced (Note: iris components handled by composite in extractDescriptions)
    'PaDNA.EyeDNA.Spacing': `${valueStr} eye spacing`,
    // Hair
    'PaDNA.HairDNA.Color': `${valueStr} hair`,
    'PaDNA.HairDNA.Length': `${valueStr} hair`,
    'PaDNA.HairDNA.Texture': `${valueStr} hair`,
    'PaDNA.HairDNA.Volume': `${valueStr} voluminous hair`,
    'PaDNA.HairDNA.Parting': `hair with ${valueStr}`,
    'PaDNA.HairDNA.Color.Base': `${valueStr} hair`,
    'PaDNA.HairDNA.Color.Highlights': `hair with ${valueStr} highlights`,
    'PaDNA.HairDNA.CurlPattern': `${valueStr.replace(/-/g, ' ')} curl pattern`,
    'PaDNA.HairDNA.Parting.Side': `${valueStr} side part`,
    // Hair - v2.1 Headpiece
    'PaDNA.HairDNA.Accessories.Headpiece': (() => {
      const headpieceMap: Record<string, string> = {
        'crown': 'wearing ornate crown',
        'tiara': 'wearing elegant tiara',
        'flower-crown': 'wearing flower crown',
        'headband': 'wearing headband',
        'veil': 'wearing veil',
        'fascinator': 'wearing fascinator',
      };
      return headpieceMap[valueStr] || null;
    })(),
    // Skin
    'PaDNA.SkinDNA.Tone': `${valueStr} skin`,
    'PaDNA.SkinDNA.Freckles': valueStr.includes('present') ? `${valueStr} freckles` : null,
    // Skin - v2.1 Texture (HIGH PRIORITY)
    'PaDNA.SkinDNA.Texture': (() => {
      const textureMap: Record<string, string> = {
        'natural-with-visible-pores': 'natural skin texture, visible pores, realistic skin detail',
        'smooth': 'smooth skin with subtle natural texture',
        'airbrushed': 'perfectly smooth skin',
        'coarse': 'textured skin with visible pores',
      };
      return textureMap[valueStr] || valueStr;
    })(),
    // Skin - Freckles (Note: individual components skipped if composite used)
    'PaDNA.SkinDNA.Undertone.Type': `${valueStr.replace(/-/g, ' ')} undertones`,
    // Face/Smile
    'PaDNA.SmileDNA.SmileShape': valueStr,
    'PaDNA.SmileDNA.Teeth': `${valueStr} teeth`,
    'PaDNA.SmileDNA.Dimples': valueStr.includes('present') ? 'dimples' : null,
    'PaDNA.SmileDNA.LipFullness': `${valueStr.replace(/-/g, ' ')} lips`,
    'PaDNA.SmileDNA.CupidsBow': `${valueStr} cupid's bow`,
    'PaDNA.FaceDNA.Cheekbones': `${valueStr} cheekbones`,
    // Expression
    'PaDNA.ExpressionDNA.TypicalExpression': valueStr,
    'PaDNA.ExpressionDNA.EyeContact': valueStr,
    // Body
    'PaDNA.BodyDNA.Build': `${valueStr} build`,
    'PaDNA.BodyDNA.WaistHipRatio': `${valueStr} figure`,
    'PaDNA.BodyDNA.Legs': `${valueStr} legs`,
    'PaDNA.BodyDNA.Arms': `${valueStr} arms`,
    'PaDNA.BodyDNA.Bust': `${valueStr} bust`,
    'PaDNA.BodyDNA.Navel': valueStr.includes('pierced') ? 'navel piercing' : null,
    // Fitness
    'PaDNA.FitnessDNA.Indicators': valueStr,
    // Apparel/Jewelry
    'PaDNA.ApparelDNA.Jewelry.Necklace': `wearing ${valueStr}`,
    'PaDNA.ApparelDNA.Jewelry.Earrings': `wearing ${valueStr}`,
    // Apparel - v2.1 Themes
    'PaDNA.ApparelDNA.Themes': (() => {
      if (!Array.isArray(value) || value.length === 0) return null;
      const themeMap: Record<string, string> = {
        'pageant': 'pageant photography style',
        'glamour': 'glamour photography',
        'beachwear': 'beach setting, casual beachwear',
        'casual': 'casual style',
        'athletic': 'athletic wear, fitness setting',
        'evening-wear': 'elegant evening wear',
        'performance': 'performance outfit',
        'editorial': 'editorial fashion photography',
      };
      return value.slice(0, 2).map((t: string) => themeMap[t] || t).join(', ');
    })(),
    // Makeup - v2.1 (MEDIUM-HIGH PRIORITY)
    'MakeupDNA.Eyes.Style': (() => {
      const styleMap: Record<string, string> = {
        'natural': 'natural eye makeup',
        'smoky-eye': 'smoky eye makeup',
        'heavy-smoky-eye-with-black-liner': 'heavy dramatic smoky eye with thick black eyeliner, intense eye makeup',
        'winged': 'winged eyeliner',
        'glam': 'glamorous eye makeup',
      };
      return styleMap[valueStr] || valueStr;
    })(),
    'MakeupDNA.Eyes.Underliner': valueStr === 'present' ? 'lower eyelid liner' : null,
    'MakeupDNA.Lashes': (() => {
      const lashMap: Record<string, string> = {
        'natural': 'natural lashes',
        'mascara': 'mascara-enhanced lashes',
        'long-volumized-mascara-false': 'long volumized lashes with dramatic mascara or false lashes',
        'false-dramatic': 'dramatic false eyelashes',
      };
      return lashMap[valueStr] || valueStr;
    })(),
    'MakeupDNA.Lips.Color': (() => {
      const colorMap: Record<string, string> = {
        'nude': 'nude lips',
        'nude-to-pink': 'nude pink lips',
        'pink': 'pink lips',
        'red': 'red lips',
        'berry': 'berry-toned lips',
      };
      return colorMap[valueStr] || `${valueStr} lips`;
    })(),
    'MakeupDNA.Lips.Finish': (() => {
      const finishMap: Record<string, string> = {
        'matte': 'matte lip finish',
        'natural': 'natural lip finish',
        'gloss': 'glossy lips, shiny finish',
        'shimmer': 'shimmery lips',
        'metallic': 'metallic lip finish',
      };
      return finishMap[valueStr] || valueStr;
    })(),
    'MakeupDNA.Skin.Finish': (() => {
      const finishMap: Record<string, string> = {
        'natural': 'natural skin finish',
        'dewy': 'dewy skin finish',
        'satin': 'satin skin finish',
        'medium-coverage-matte': 'medium-coverage matte foundation',
        'full-coverage-matte': 'full-coverage matte foundation',
      };
      return finishMap[valueStr] || valueStr;
    })(),
    'MakeupDNA.Blush.Tone': `${valueStr} blush`,
  };

  const desc = descriptions[path];

  // Fallback: use value directly
  if (desc === undefined && value) {
    return valueStr;
  }

  return desc ?? null;
}

function extractDescriptions(traits: TraitDict, emphasisThreshold: number): TraitDescription {
  const descriptions: TraitDescription = {};

  // v2.1: Track composite components
  const irisComponents = {
    base: traits['PaDNA.EyeDNA.Iris.BaseColor']?.resolved_value,
    flecks: traits['PaDNA.EyeDNA.Iris.Flecks']?.resolved_value,
    pattern: traits['PaDNA.EyeDNA.Iris.FleckPattern']?.resolved_value,
    ring: traits['PaDNA.EyeDNA.Iris.LimbalRing']?.resolved_value,
  };

  const frecklesComponents = {
    density: traits['PaDNA.SkinDNA.Freckles.Density']?.resolved_value,
    coverage: traits['PaDNA.SkinDNA.Freckles.Coverage']?.resolved_value,
    contrast: traits['PaDNA.SkinDNA.Freckles.Contrast']?.resolved_value,
  };

  // v2.1: Build sorted trait list by priority × UCN
  const sortedTraits = Object.entries(traits).map(([path, trait]) => {
    const priority = PRIORITY_WEIGHTS[path] || 'std';
    const ucn = trait?.ucn ?? 800;
    const weight = WEIGHT_VALUES[priority] * ucn;
    return { path, trait, weight };
  }).sort((a, b) => b.weight - a.weight);

  // v2.1: Add iris composite first if components exist
  if (irisComponents.base) {
    const irisDesc = irisComposite(
      irisComponents.base,
      irisComponents.flecks,
      irisComponents.pattern,
      irisComponents.ring
    );
    if (irisDesc) {
      if (!descriptions.eyes) descriptions.eyes = [];
      descriptions.eyes.push(irisDesc);
    }
  }

  // v2.1: Add freckles composite first if components exist
  if (frecklesComponents.density) {
    const frecklesDesc = frecklesComposite(
      frecklesComponents.density,
      frecklesComponents.coverage,
      frecklesComponents.contrast
    );
    if (frecklesDesc) {
      if (!descriptions.skin) descriptions.skin = [];
      descriptions.skin.push(frecklesDesc);
    }
  }

  // Skip individual iris/freckles components (handled by composites)
  const skipPaths = new Set([
    'PaDNA.EyeDNA.Iris.BaseColor',
    'PaDNA.EyeDNA.Iris.Flecks',
    'PaDNA.EyeDNA.Iris.FleckPattern',
    'PaDNA.EyeDNA.Iris.LimbalRing',
    'PaDNA.SkinDNA.Freckles.Density',
    'PaDNA.SkinDNA.Freckles.Coverage',
    'PaDNA.SkinDNA.Freckles.Contrast',
  ]);

  for (const { path, trait } of sortedTraits) {
    if (!trait || typeof trait !== 'object') continue;
    if (skipPaths.has(path)) continue;

    const value = trait.resolved_value;
    if (value === null || value === undefined) continue;

    const category = categorizeTraitPath(path);
    if (!descriptions[category]) {
      descriptions[category] = [];
    }

    const desc = traitToDescription(path, value);
    if (desc) {
      descriptions[category].push(desc);
    }
  }

  return descriptions;
}

/**
 * Build an optimized prompt from PaDNA traits
 */
export function buildPromptFromTraits(
  traits: TraitDict,
  style: PromptStyle = 'photorealistic',
  emphasisThreshold: number = 800.0,
  maxTokens: number = 200
): string {
  const segments: string[] = [];

  // Style prefix
  segments.push(STYLE_PREFIXES[style]);

  // Build descriptions by category
  const descriptions = extractDescriptions(traits, emphasisThreshold);

  // Collect face parts
  const faceParts: string[] = [];

  // Eyes
  if (descriptions.eyes) {
    faceParts.push(...descriptions.eyes);
  }

  // Hair
  if (descriptions.hair) {
    faceParts.push(...descriptions.hair);
  }

  // Skin
  if (descriptions.skin) {
    faceParts.push(...descriptions.skin);
  }

  // Face features and smile
  if (descriptions.smile) {
    faceParts.push(...descriptions.smile);
  }
  if (descriptions.face) {
    faceParts.push(...descriptions.face);
  }

  // Core subject with face features
  if (faceParts.length > 0) {
    segments.push(`a woman with ${faceParts.join(', ')}.`);
  } else {
    segments.push('a woman.');
  }

  // v2.1: Makeup (medium-high priority)
  if (descriptions.makeup && descriptions.makeup.length > 0) {
    segments.push(descriptions.makeup.join(', ') + '.');
  }

  // Expression
  if (descriptions.expression && descriptions.expression.length > 0) {
    segments.push(`Expression: ${descriptions.expression.slice(0, 2).join(', ')}.`);
  }

  // Body and fitness
  const bodyParts: string[] = [];
  if (descriptions.body) {
    bodyParts.push(...descriptions.body.slice(0, 4));
  }
  if (descriptions.fitness) {
    bodyParts.push(...descriptions.fitness.slice(0, 2));
  }

  if (bodyParts.length > 0) {
    segments.push(`Body: ${bodyParts.join(', ')}.`);
  }

  // Details (jewelry, apparel)
  if (descriptions.apparel && descriptions.apparel.length > 0) {
    const details = descriptions.apparel.slice(0, 3);
    segments.push(`Wearing: ${details.join(', ')}.`);
  }

  // Join segments
  let prompt = segments.join(' ');

  // Add quality keywords
  prompt += ', sharp focus, high detail, natural lighting';

  // Truncate if too long
  const words = prompt.split(/\s+/);
  if (words.length > maxTokens) {
    prompt = words.slice(0, maxTokens).join(' ') + '...';
  }

  return prompt;
}

/**
 * Get default negative prompt (v2.1 enhanced)
 */
export function getDefaultNegativePrompt(traits?: TraitDict): string {
  let negativePrompt =
    'blurry, low quality, distorted, deformed, ugly, bad anatomy, ' +
    'bad proportions, extra limbs, mutated, disfigured, ' +
    'watermark, text, signature, logo, multiple heads, duplicate, ' +
    'cartoon, anime, painting, illustration';

  // v2.1: Add anti-airbrushing when natural texture is specified
  if (traits) {
    const skinTexture = traits['PaDNA.SkinDNA.Texture']?.resolved_value;
    if (skinTexture === 'natural-with-visible-pores') {
      negativePrompt += ', NOT airbrushed, NOT overly smooth skin, NOT plastic-looking';
    }
  }

  return negativePrompt;
}
